### DISCUSSION TITLE
State of the TypeScript migration

### DISCUSSION THREAD

**Discussion by nunomaduro** [algolia/instantsearch#4407] [General]
Hey team, the goal of this discussion is share thoughts of the TypeScript migration.

Jira Task:  [IFW-1152](https://algolia.atlassian.net/secure/RapidBoard.jspa?rapidView=10&projectKey=IFW&modal=detail&selectedIssue=IFW-1152)
This need to reviewed, tested, and merged: https://github.com/algolia/instantsearch.js/pull/4408.
Current content migrated: **45%**. The more complicated connectors/widgets are done.

The amount of time needed to migrated the remaining **55%** it may take 4 weeks of work. It's easy task, but very time consuming to do it right.

Note that, this migration is focus on the public API of connectors/widgets, and not on making internals strict.

---

## How to migrate a connector/widget 

1. Connector & Connector Tests

First, you migrate the **entire** connector folder. Keep in mind,
that for each connector you need to export:

```
- The default function with the name: **connect{Name}** and his type **{Name}Connector**.
- The **{Name}RendererOptions** type and its used types.
- The **{Name}ConnectorParams** type and its used types.
```

2. Widget & Widgets Tests

Next, you migrate the **entire** widget folder. Keep in mind,
that for each widget you need to export:

```
- The default function with the name: **{name}** and his type **{Name}Widget**.
- The **{Name}WidgetOptions** type, and its used types.
- Optionally the **{Name}Templates** type and its used types.
- Optionally the **{Name}CSSClasses** type and its used types.
```

3. Stories

Finally, you need to upgrade the stories to used the typed
widget/connectors replacing:

```
+ import { {name} } from '../../src/widgets';

search.addWidgets([
-  instantsearch.widgets.{name}({ ..
+  {name}({ ..
```

**Comment by francoischalifour** [comment]
Thanks for bootstrapping this task.

To add on the migration process: if you feel adventurous, you can also migrate the Preact components 🙂
